File: manual_formula/py_src/doc_reader.py
import regex as re
from math import log
import glob

r1 = re.compile(r"[^\p{L}]")

class DocAll:
    def __init__(self, sws, g_abstr):
        self.data = dict()
        self.no_documents = 0

        for entry in glob.glob(g_abstr):
            self.no_documents += 1
            with open(entry, 'r') as f:
                text = f.read().lower()
            cset = set([x for x in re.split(r1, text) if not x in sws])

            for word in cset:
                if not word in self.data:
                    self.data[word] = 0
                self.data[word] += 1

    def term_idf(self, term):
        total = self.data.get(term, 1)
        return log(self.no_documents, 10.0) / total

File: cs_src/DocReader.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;

namespace smake
{
    class DocAll
    {
        public Dictionary<string, int> Data { get; private set; }
        public int DocsNum { get; private set; }

        public DocAll(HashSet<string> sws)
        {
            var data = new Dictionary<string, int>();
            int docsNum = 0;

            foreach (var entry in Directory.GetFiles(Options.G_ABSTR_DIR, Options.G_ABSTR_PATTERN))
            {
                docsNum += 1;
                var text = File.ReadAllText(entry).ToLower();
                var textWords = Regex.Split(text, @"[^\p{L}]");
                var cset = new HashSet<string>(textWords.Where(x => !sws.Contains(x)));

                foreach (var word in cset)
                {
                    data.Add1OrIncrement(word);
                }
            }

            this.Data = data;
            this.DocsNum = docsNum;
        }

        public double TermIdf(string term)
        {
            var total = this.Data.GetValueOrDefault(term, 1);
            return Math.Log(this.DocsNum, 10.0) / total;
        }

    }
}
